### PR TITLE
[TA] Add encoding support and length and offset

### LINKED ISSUE #10710: Add support for parameter `string-index-type`
We want to keep this parameter hidden from the user, and the idea right now is that we are going to calculate the encoding and send it to the service.
Same, we are not going to expose it in the result objects, although we should have it available for logging for diagnosability purposes.

Important add documentation about this in the readme, samples, and docstrings to make sure the user knows what is happening.

### LINKED ISSUE #14552: Add length and offset properties to all endpoints

### PR DESCRIPTION
**Pull request by maririos**
Fixes: https://github.com/Azure/azure-sdk-for-net/issues/10710 - Support for encoding. For C# default is `Utf16CodeUnit`
Fixes: https://github.com/Azure/azure-sdk-for-net/issues/14552 - Brings back `Length` and `Offset`

Updates samples, tests, and adds tests for encoding

### CODE REVIEW

**Review by mssfang — APPROVED** [review]

File: sdk/textanalytics/Azure.AI.TextAnalytics/src/TextAnalyticsClient.cs [hunk 1]
```diff
@@ -26,6 +26,10 @@ public partial class TextAnalyticsClient
         private readonly string DefaultCognitiveScope = "https://cognitiveservices.azure.com/.default";
         private const string AuthorizationHeader = "Ocp-Apim-Subscription-Key";
 
+        // Specifies the method used to interpret string offsets. Default to <see cref="StringIndexType.Utf16CodeUnit"/>.
+        private readonly StringIndexType _stringCodeUnit = StringIndexType.Utf16CodeUnit;
```
Comment (kinelski) [inline sdk/textanalytics/Azure.AI.TextAnalytics/src/TextAnalyticsClient.cs:30]: Just out of curiosity, do we intend to let users overwrite this value?

**Review by kinelski — APPROVED** [review]
LGTM.

File: sdk/textanalytics/Azure.AI.TextAnalytics/CHANGELOG.md [hunk 2]
```diff
@@ -6,6 +6,7 @@
 - It defaults to the latest supported API version, which currently is `3.1-preview.2`.
 - `ErrorCode` value returned from the service is now surfaced in `RequestFailedException`.
 - Support added for Opinion Mining. This feature is available in the Text Analytics service v3.1-preview.1 and above.
+- Added `Offset` and `Length` properties for `CategorizedEntity`, `SentenceSentiment`, and `LinkedEntityMatch`. The default encoding is UTF16 code units. For additional information see https://aka.ms/text-analytics-offsets
```
[the remark this answers is not in the document]
Comment (maririos) [reply sdk/textanalytics/Azure.AI.TextAnalytics/CHANGELOG.md:9]: Yeah. I thought of not including it here in the changelog as `AspectSentiment` and `OpinionSentiment` are new concepts for the user, so everything there is going to be new

File: sdk/textanalytics/Azure.AI.TextAnalytics/tests/RecognizeLinkedEntitiesTests.cs [hunk 3]
```diff
@@ -53,8 +53,13 @@ public async Task RecognizeLinkedEntitiesTest()
...
                 Assert.IsNotNull(entity.Url);
                 Assert.IsNotNull(entity.Matches);
-                Assert.IsNotNull(entity.Matches.First().ConfidenceScore);
-                Assert.IsNotNull(entity.Matches.First().Text);
+
+                LinkedEntityMatch match = entity.Matches.First();
+                Assert.IsNotNull(match.ConfidenceScore);
+                Assert.IsNotNull(match.Text);
+                Assert.IsNotNull(match.Offset);
+                Assert.IsNotNull(match.Length);
```
[the remark this answers is not in the document]
Comment (maririos) [reply sdk/textanalytics/Azure.AI.TextAnalytics/tests/RecognizeLinkedEntitiesTests.cs:61]: hmm great point. the only one that makes sense is for length, so I will remove the `IsNotNull` checks

File: sdk/textanalytics/Azure.AI.TextAnalytics/src/TextAnalyticsClient.cs (hunk 1 above)
Comment (maririos) [reply sdk/textanalytics/Azure.AI.TextAnalytics/src/TextAnalyticsClient.cs:30]: Not currently. There was some talking about the service returning 2 Offset and Length with different encoding and us exposing them, but for now only the one the language allows (UTF-16 for C#).

File: sdk/textanalytics/Azure.AI.TextAnalytics/README.md [hunk 4]
```diff
@@ -190,7 +190,8 @@ CategorizedEntityCollection entities = client.RecognizeEntities(document);
 Console.WriteLine($"Recognized {entities.Count} entities:");
 foreach (CategorizedEntity entity in entities)
 {
-    Console.WriteLine($"Text: {entity.Text}, Category: {entity.Category}, SubCategory: {entity.SubCategory}, Confidence score: {entity.ConfidenceScore}");
+    Console.WriteLine($"Text: {entity.Text}, Offset (in UTF-16  code units): {entity.Offset}, Length (in UTF-16  code units): {entity.Length}");
```
[the remark this answers is not in the document]
Comment (maririos) [reply sdk/textanalytics/Azure.AI.TextAnalytics/README.md]: Nice, thank you! updated

File: sdk/textanalytics/Azure.AI.TextAnalytics/api/Azure.AI.TextAnalytics.netstandard2.0.cs [hunk 5]
```diff
@@ -22,6 +22,8 @@ public partial class AnalyzeSentimentResultCollection : System.Collections.Objec
         private readonly object _dummy;
         private readonly int _dummyPrimitive;
         public Azure.AI.TextAnalytics.SentimentConfidenceScores ConfidenceScores { get { throw null; } }
+        public int Length { get { throw null; } }
+        public int Offset { get { throw null; } }
```
Comment (tg-msft) [inline sdk/textanalytics/Azure.AI.TextAnalytics/api/Azure.AI.TextAnalytics.netstandard2.0.cs:26]: How does `Length` relate to `Text.Length`?

It'd be worth discussing in ApiView if there's a prefix we can use for `Offset` to make it clearer to new users of the service.  I also think .NET more commonly calls these things `Index` (like https://docs.microsoft.com/en-us/dotnet/api/system.text.regularexpressions.capture.index?view=netcore-3.1#System_Text_RegularExpressions_Capture_Index) but that might be kind of dated now that we have System.Index.

Comment (maririos) [reply sdk/textanalytics/Azure.AI.TextAnalytics/api/Azure.AI.TextAnalytics.netstandard2.0.cs:26]: >How does Length relate to Text.Length?

There is no difference. That is a great point :| 
I will add both comments into the API View

**Pull request merged**